pkg/api/latest/latest.go

```diff
@@ -76,6 +80,20 @@ func (g GroupMetaMap) GroupOrDie(group string) *GroupMeta {
 	return groupMeta
 }
 
+// AllPreferredGroupVersions returns the preferred versions of all registered
+// groups in the form of "group1/version1,group2/version2,..."
+func (g GroupMetaMap) AllPreferredGroupVersions() string {
+	if len(g) == 0 {
+		return ""
+	}
+	var defaults string
+	for _, groupMeta := range g {
+		defaults += groupMeta.GroupVersion + ","
+	}
+	defaults = strings.TrimRight(defaults, ",")
```

Comment:
The sort is nice for users, too-- it'd be weird if the default changed
every time they ran help!

(Also if you're worried about performance--which doesn't matter here--you
should construct everything in a []byte--string concatenation is expensive
in go.)

On Wed, Sep 23, 2015 at 5:56 PM, Chao Xu <notifications@github.com> wrote:

> In pkg/api/latest/latest.go
> <https://github.com/kubernetes/kubernetes/pull/14397#discussion_r40276293>
> :
>
> > @@ -76,6 +80,20 @@ func (g GroupMetaMap) GroupOrDie(group string) *GroupMeta {
> >  	return groupMeta
> >  }
> >
> > +// AllPreferredGroupVersions returns the preferred versions of all registered
> > +// groups in the form of "group1/version1,group2/version2,..."
> > +func (g GroupMetaMap) AllPreferredGroupVersions() string {
> > +	if len(g) == 0 {
> > +		return ""
> > +	}
> > +	var defaults string
> > +	for _, groupMeta := range g {
> > +		defaults += groupMeta.GroupVersion + ","
> > +	}
> > +	defaults = strings.TrimRight(defaults, ",")
>
> Sure. I'm just trying to improve the efficiency by skipping the sort, as
> we don't need a deterministic order except for test.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/14397/files#r40276293>.
>